pontoon/base/views.py

```diff
@@ -364,20 +367,80 @@ def contributors(request, template='users.html'):
         .exclude(string=F('entity__string_plural'))
     )
 
+    translations_total = (
+        translations.annotate(count=Count('user'))
+    )
+
+    translations_approved = (
+        translations.filter(approved=True).annotate(count=Count('user'))
+    )
+
+    translations_unapproved = (
+        translations.exclude(approved=True).exclude(fuzzy=True).annotate(count=Count('user'))
+    )
+
+    translations_needs_work = (
+        translations.filter(fuzzy=True).annotate(count=Count('user'))
+    )
+
     users = (
         User.objects
         .filter(translation__in=translations).distinct()
         .exclude(email__in=settings.EXCLUDE)
         .annotate(translation_count=Count('translation'))
         .exclude(translation_count=0)
-        .order_by('-translation_count')[:100]
     )
 
-    for user in users:
-        user.translations = translations.filter(user=user)
+    today = datetime.date.today()
+
+    end_date = datetime.datetime(today.year, today.month, 1)
+    start_date = None
+
+    if period == '12months':
+        start_date = end_date - relativedelta(month=+12)
+    elif period == '6months':
+        start_date = end_date - relativedelta(month=+6)
+    elif period == '3months':
+        start_date = end_date - relativedelta(month=+3)
+    elif period == 'lastmonth':
+        start_date = end_date - relativedelta(month=+1)
+
+    if start_date and end_date:
+        users = users.filter(translation__date__range=(start_date, end_date))
+        translations_total = (
+            translations_total.filter(date__range=(start_date, end_date))
+        )
+        translations_approved = (
+            translations_approved.filter(date__range=(start_date, end_date))
+        )
+        translations_unapproved = (
+            translations_unapproved.filter(date__range=(start_date, end_date))
+        )
+        translations_needs_work = (
+            translations_needs_work.filter(date__range=(start_date, end_date))
+        )
+
+    def translation_counts(translations):
+        counts = defaultdict(int)
+        for user_id in translations.values_list('user__pk', flat=True):
+            counts[user_id] += 1
+        return counts
+
+    translations_total_map = translation_counts(translations_total)
+    translations_approved_map = translation_counts(translations_approved)
+    translations_unapproved_map = translation_counts(translations_unapproved)
+    translations_needs_work_map = translation_counts(translations_needs_work)
+    contributors = list(users.order_by('-translation_count')[:100])
+
+    for user in contributors:
+        user.translations_total = translations_total_map.get(user.pk, 0)
+        user.translations_approved = translations_approved_map.get(user.pk, 0)
+        user.translations_unapproved = translations_unapproved_map.get(user.pk, 0)
+        user.translations_needs_work = translations_needs_work_map.get(user.pk, 0)
```

Comment:
A good rule of thumb is that if you're ever manually assigning values to an instance of a class, you should consider if it instead could be a property on that class. In this case, it makes a whole lot of sense to just make `User.translation_count`, `User.approved_translation_count`, etc. as properties on the `User` class.

(There's some other options that actually might be better, like a custom translation manager, but those are kinda complex and have some nasty edge cases that we probably just don't want to bother with.)

This is probably a good time for us to set up a [custom User model](https://docs.djangoproject.com/en/1.8/topics/auth/customizing/#specifying-a-custom-user-model) for this and other purposes.